Fix numeric keysym parsing

wismill opened the pull request:
While working on #332, I realized that numeric keysyms parsing has some issues. Let’s deal with it in this separate PR.

This PR fixes some edges cases and improve tests.

bluetech commented on include/xkbcommon/xkbcommon-keysyms.h:
```diff
@@ -3250,6 +3250,11 @@ performance, or use of this material.
+ * Keysyms are 32-bit integers with the 3 most significant bits always set
+ * to zero. See: Appendix A “KEYSYM Encoding” of the X Window System Protocol.
+ */
+#define XKB_MAX_KEYSYM               0x1fffffff
```
This define should go in `xkbcommon.h`, like the other MAX defines. For consistency, it should be called `XKB_KEYSYM_MAX`.

wismill replied:
Done

bluetech commented on src/keysym.c:
```diff
@@ -64,7 +64,7 @@ get_name(const struct name_keysym *entry)
 xkb_keysym_get_name(xkb_keysym_t ks, char *buffer, size_t size)
 {
-    if ((ks & ((unsigned long) ~0x1fffffff)) != 0) {
+    if ((ks & ((unsigned long) ~XKB_MAX_KEYSYM)) != 0) {
```
This looks a bit weird, e.g. if the max was 535353 then it wouldn't have made sense to apply `~` to it. With a MAX constant it's clearer to do `if (ks > XKB_MAX_KEYSYM) { ... }`.

wismill replied:
Yes. I went the lazy road to keep diff minimal.

bluetech commented on src/xkbcomp/expr.c:
```diff
@@ -656,11 +656,25 @@ ExprResolveKeySym(struct xkb_context *ctx, const ExprDef *expr,
     if (!ExprResolveInteger(ctx, expr, &val))
         return false;
 
-    if (val < 0 || val >= 10)
```
This looks like it regressed in 2e4933cda7c141c9efc8c62e7d72f591f2998836 compared to xkbcomp...

bluetech commented on test/keysym.c:
```diff
@@ -27,6 +27,9 @@
 #include "test.h"
 #include "keysym.h" /* For unexported is_lower/upper/keypad() */
 
+#define STRINGIFY_(x) #x
```
Maybe move this to utils.h?

Also, maybe change `STRINGIFY_` -> `STRINGIFY` and `STRINGIFY` -> `STRINGIFY2`? Sometimes you don't want to expand (e.g. the one in `x11/keymap.c` doesn't.

wismill replied:
Done

bluetech reviewed (commented):
Good fixes!

I left some comments. Also, do you mind splitting the `Fix parsing of numeric keysyms` commit such that each fix is in its own commit? It will be easier to bisect etc. this way.

wismill commented:
@bluetech Commit split!

bluetech reviewed (approved):
Thanks!

whot commented on include/xkbcommon/xkbcommon.h:
```diff
@@ -303,6 +303,14 @@ typedef uint32_t xkb_led_mask_t;
+ *
+ * @since 1.6.0
+ */
+#define XKB_KEYSYM_MAX      (0x1fffffff)
```
aside from consistency with the ones above but - why is this in parenthesis? 

`XKB_KEYCODE_MAX` needs it (though it would be trivial to just specify it as the actual numeric value instead with a comment for the humand-readable one but here? I'm struggling to think of where the parens make a difference.

wismill replied:
I used parentheses just for consistency with the other ones. Since the new constants has no special requirement, let’s drop them.

whot commented on test/keysym.c:
```diff
@@ -27,6 +27,9 @@
 #include "keysym.h" /* For unexported is_lower/upper/keypad() */
 
+/* Needed because XKB_KEYSYM_MAX has parentheses */
+#define KEYSYM_MAX 0x1fffffff
```
tbh I think it'd be better off using the actual strings in the tests instead of having a separate `#define` here. e.g
```c
    assert(test_string("0x1fffffff", XKB_KEYSYM_MAX));
```
is immediately understandable without the extra redirection.

wismill replied:
I dropped the parentheses so we can use XKB_KEYSYM_MAX for clarity.

whot reviewed (approved):
two nitpicks but this LGTM. though my eyes glaze over every time i look at the actual parser code :)

wismill commented:
Rebased.

I added the constant `XKB_KEYSYM_MIN`. `xkb_keysym_t` is unsigned, but we do need to check min keysym in parser. Then we have the couple min/max.

I also added much more test related to keysyms, to reflect the current implementation.

wismill commented on include/xkbcommon/xkbcommon.h:
```diff
@@ -303,6 +312,24 @@ typedef uint32_t xkb_led_mask_t;
+ * @sa XKB_KEYSYM_MAX, xkb_keysym_t
+ * @ingroup keysyms
+ */
+#define XKB_KEYSYM_MIN      0x00000000
```
Good point
